nemo/collections/nlp/models/language_modeling/megatron_ptune_t5_model.py

```diff
@@ -0,0 +1,494 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List
+
+import torch
+from omegaconf import OmegaConf, open_dict
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.data.glue_benchmark.t5_ptune_dataset import T5PTuneDataset, T5PTuneInferenceDataset
+from nemo.collections.nlp.models.language_modeling.megatron_t5_model import MegatronT5Model
+from nemo.collections.nlp.models.nlp_model import NLPModel
+from nemo.collections.nlp.modules.common.megatron.megatron_init import initialize_model_parallel_for_nemo
+from nemo.collections.nlp.modules.common.megatron.utils import (
+    average_losses_across_data_parallel_group,
+    build_position_ids,
+)
+from nemo.collections.nlp.modules.common.prompt_encoder import PromptEncoder
+from nemo.utils import logging
+
+try:
+    from apex.transformer import tensor_parallel
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ['MegatronT5PTuneModel']
+
+
+class MegatronT5PTuneModel(NLPModel):
+    """
+    Megatron T5 P-Tune
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+        initialize_model_parallel_for_nemo(
+            world_size=trainer.world_size,
+            global_rank=trainer.global_rank,
+            local_rank=trainer.local_rank,
+            tensor_model_parallel_size=cfg.get('tensor_model_parallel_size', 1),
+            seed=cfg.get('seed', 1234),
+        )
+
+        # shared params for dataset and data loaders
+        # tokenizer needs to get initialized before the super.__init__()
+        # as dataloaders and datasets need it to process the data
+
+        self.megatron_amp_o2 = cfg.get('megatron_amp_O2', False)
```

Comment:
Add this initialization to `MegatronBaseModel` ?

Reply:
This is a workaround for the apex bug (https://github.com/NVIDIA/apex/issues/1311). We will remove it once the bug is fixed. No need to put it in the base.